osu.Game/Online/Multiplayer/StatefulMultiplayerClient.cs

```diff
@@ -109,30 +108,66 @@ protected StatefulMultiplayerClient()
             });
         }
 
+        private readonly object joinOrLeaveTaskLock = new object();
+        private Task? joinOrLeaveTask;
+
         /// <summary>
         /// Joins the <see cref="MultiplayerRoom"/> for a given API <see cref="Room"/>.
         /// </summary>
         /// <param name="room">The API <see cref="Room"/>.</param>
         public async Task JoinRoom(Room room)
         {
-            if (Room != null)
-                throw new InvalidOperationException("Cannot join a multiplayer room while already in one.");
+            Task? lastTask;
+            Task newTask;
 
-            Debug.Assert(room.RoomID.Value != null);
+            lock (joinOrLeaveTaskLock)
+            {
+                lastTask = joinOrLeaveTask;
+                joinOrLeaveTask = newTask = Task.Run(async () =>
+                {
+                    if (lastTask != null)
+                        await lastTask;
 
-            apiRoom = room;
-            playlistItemId = room.Playlist.SingleOrDefault()?.ID ?? 0;
+                    // Should be thread-safe since joinOrLeaveTask is locked on in both JoinRoom() and LeaveRoom().
+                    if (Room != null)
+                        throw new InvalidOperationException("Cannot join a multiplayer room while already in one.");
```

Comment:
I'm going to be honest, I'm nowhere near being even 10% sure this is thread-safe given the current complexity of this code.

All current mutations to `Room` are done inside of a schedule inside of a TPL task, and the task to which the field is assigned to is protected by a lock. Which by itself is a mouthful and a half, but the TPL part and schedule part is especially why I'm skeptical this is safe.

As far as I can see the only part that the lock protects is the assignment to `joinOrLeaveTask` - IOW, it makes sure that a leave task can't be "dropped" by a new join and not awaited on, due to it being written over, and vice-versa.

It does *not*, as far as I can see, guarantee that `lastTask` is not running concurrently alongside `newTask`, because the *body* of `Task.Run()` is not in the critical section (and even if it was, it'd still be super ultra bad, because awaits inside critical sections sound like immediate deadlock trouble, and should really fail to compile if I'm not mistaken). In theory:

```
old leave task			new join task

						reads non-null room

executes scheduled portion
clears out room

						continues past check and explodes
```

Am I missing something?

Reply:
Am I wrong in thinking that it's the `await lastTask` above which guarantees that it's not running alongside `newTask`?

Reply:
I don't think so, as I believe `Task.Run()` will kick off the task potentially immediately? I'm thinking the await on the old task won't help if you don't delay starting the new task after it, but I haven't experimentally confirmed this.

Reply:
Never mind, I forgot that the await is inside the tasks. It still feels a little dicey though.